src/main/java/eu/okaeri/validator/OkaeriValidator.java

```diff
@@ -93,9 +101,79 @@ public Set<ConstraintViolation> validateParameters(@NonNull Parameter[] paramete
-        return this.validationProviders.values().stream()
+        Set<ConstraintViolation> violations = this.validationProviders.values().stream()
             .filter(provider -> provider.shouldValidate(parameter))
             .flatMap(provider -> provider.validate(parameter, value).stream())
             .collect(Collectors.toCollection(LinkedHashSet::new));
+        violations.addAll(this.validate(parameter.getName(), parameter.getAnnotatedType(), value));
+        return violations;
+    }
+
+    private Set<ConstraintViolation> validate(@NonNull String fieldName, @NonNull AnnotatedType annotatedObjectType, Object object) {
+        if (!(annotatedObjectType instanceof AnnotatedParameterizedType)) {
+            return new LinkedHashSet<>();
+        }
+        AnnotatedParameterizedType parameterizedType = (AnnotatedParameterizedType) annotatedObjectType;
+
+        if (!(parameterizedType.getType() instanceof ParameterizedType)) {
+            return new LinkedHashSet<>();
+        }
+        Class<?> objectType = (Class<?>) ((ParameterizedType) parameterizedType.getType()).getRawType();
+
+        AnnotatedType[] actualTypes = parameterizedType.getAnnotatedActualTypeArguments();
+        if (Collection.class.isAssignableFrom(objectType)) {
+            return this.validateCollection(fieldName, actualTypes[0], (Collection<?>) object);
+        } else if (Map.class.isAssignableFrom(objectType)) {
+            return this.validateMap(fieldName, actualTypes[0], actualTypes[1], (Map<?, ?>) object);
+        }
+        return new LinkedHashSet<>();
+    }
+
+    @Override
+    public Set<ConstraintViolation> validateCollection(@NonNull String fieldName, @NonNull AnnotatedType annotatedValueType, @NonNull Collection<?> collection) {
+        Type valueType = annotatedValueType.getType();
+        if (!(valueType instanceof Class<?>)) {
+            return new LinkedHashSet<>();
+        }
+
+        AtomicInteger index = new AtomicInteger(0);
+        return collection.stream()
+            .flatMap(element -> this.validateElement(fieldName, (Class<?>) valueType, annotatedValueType, element).stream()
+                .peek(elementViolation -> elementViolation.setField(fieldName + " [element index " + index.getAndIncrement() + "]")))
+            .collect(Collectors.toCollection(LinkedHashSet::new));
+    }
+
+    @Override
+    public Set<ConstraintViolation> validateMap(@NonNull String fieldName, @NonNull AnnotatedType annotatedKeyType, @NonNull AnnotatedType annotatedValueType, @NonNull Map<?, ?> map) {
+        Type keyType = annotatedKeyType.getType();
+        Type valueType = annotatedValueType.getType();
+        if (!(keyType instanceof Class<?>) || !(valueType instanceof Class<?>)) {
+            return new LinkedHashSet<>();
+        }
+
+        return map.entrySet().stream()
+            .flatMap(entry -> {
+                Set<ConstraintViolation> entryViolations = new LinkedHashSet<>();
```

Comment:
I think Stream.concat on both validateElement...collect would produce much cleaner code and allow inline lambda.